src/Symfony/Component/CssSelector/Tests/XPath/TranslatorTest.php

```diff
@@ -219,6 +219,11 @@ public static function getCssToXPathTestData()
             ['e + f', "e/following-sibling::*[(name() = 'f') and (position() = 1)]"],
             ['e ~ f', 'e/following-sibling::f'],
             ['div#container p', "div[@id = 'container']/descendant-or-self::*/p"],
+            ['div:has(div.foo)', "div[count(descendant-or-self::div[@class and contains(concat(' ', normalize-space(@class), ' '), ' foo ')]) > 0]"],
+            ['div:has(.foo .bar)', "div[count(descendant-or-self::*[@class and contains(concat(' ', normalize-space(@class), ' '), ' foo ')]/descendant-or-self::*/*[@class and contains(concat(' ', normalize-space(@class), ' '), ' bar ')]) > 0]"],
+            ['div:has(> .foo)', "div[count(descendant-or-self::*/*[@class and contains(concat(' ', normalize-space(@class), ' '), ' foo ')]) > 0]"],
+            ['div:has(~ .foo)', "div[count(descendant-or-self::*/following-sibling::*[@class and contains(concat(' ', normalize-space(@class), ' '), ' foo ')]) > 0]"],
```

Comment:
`descendant-or-self::*` looks wrong for those combinators. I'm quite sure it should be `self::*` instead

Reply:
My guess is that you currently parse `div:has(> .foo)` as `div:has(* > .foo)`, which is totally different.